src/PackagePublisher/Program.cs

```
@@ -126,65 +126,75 @@ private static bool AreValidParameters()
         {
-            using (var semaphore = new SemaphoreSlim(initialCount: 4, maxCount: 4))
-            {
-                var sourceRepository = Repository.Factory.GetCoreV3(_feedToUploadTo.Value(), FeedType.HttpV3);
-                var metadataResource = await sourceRepository.GetResourceAsync<MetadataResource>();
-                var packageUpdateResource = await sourceRepository.GetResourceAsync<PackageUpdateResource>();
+            var sourceRepository = Repository.Factory.GetCoreV3(_feedToUploadTo.Value(), FeedType.HttpV3);
+            var metadataResource = await sourceRepository.GetResourceAsync<MetadataResource>();
+            var packageUpdateResource = await sourceRepository.GetResourceAsync<PackageUpdateResource>();
+
+            var tasks = Enumerable
+                .Range(0, _maxParallelPackagePushes)
+                .Select(_ => PushPackagesAsync(packageUpdateResource));
+
+            await Task.WhenAll(tasks);
+        }
+
+        private static async Task PushPackagesAsync(PackageUpdateResource packageUpdateResource)
+        {
+            PackageInfo package = null;
+            var exceptions = new List<Exception>();
 
-                var tasks = packages.Select(async package =>
+            while (_packageQueue.TryDequeue(out package))
+            {
+                try
                 {
-                    await semaphore.WaitAsync(TimeSpan.FromMinutes(3));
-                    try
-                    {
-                        await PushPackageAsync(packageUpdateResource, package);
-                    }
-                    finally
-                    {
-                        semaphore.Release();
-                    }
-                });
-                await Task.WhenAll(tasks);
+                    await PushPackageAsync(packageUpdateResource, package);
+                }
+                catch (Exception ex)
+                {
+                    // do not block from the next package to be pushed
```

Sounds like @pranavkm agreed that fail fast is fine.